src/basic/path-util.c

```diff
@@ -703,47 +721,64 @@ int find_executable_full(const char *name, const char *root, bool use_path_envva
                 if (r == 0)
                         break;
 
-                if (!path_is_absolute(element))
-                        continue;
+                if (find_executable_from_path(name, root, element, &last_error, ret_filename, ret_fd, fd))
+                        return 0;
+        }
 
-                if (!path_extend(&element, name))
-                        return -ENOMEM;
+        return last_error;
+}
 
-                if (root) {
-                        char *path_name;
+bool find_executable_from_path(
+        const char *name,
+        const char *root,
+        const char *element,
+        int *last_error_addr,
+        char **ret_filename,
+        int *ret_fd,
+        int fd) {
+        int r;
+        char *dup_element = strdup(element);
+        if (!path_is_absolute(dup_element))
+                return false;
 
-                        r = chase_symlinks(element,
-                                           root,
-                                           CHASE_PREFIX_ROOT,
-                                           &path_name,
-                                           /* ret_fd= */ NULL);
-                        if (r < 0) {
-                                if (r != -EACCES)
-                                        last_error = r;
-                                continue;
-                        }
+        if (!path_extend(&dup_element, name))
+                return false;
 
-                        free_and_replace(element, path_name);
-                }
+        if (root) {
+                char *path_name;
 
-                r = check_x_access(element, ret_fd ? &fd : NULL);
+                r = chase_symlinks(
```

Comment:
On the main branch there are were a bunch of comments here that seem to have been dropped. Could you put them back?